pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,475 @@
+	///  * mint_into (2 reads, 2 writes): Account, TotalIssuance
+	///  * deposit_stake (4 reads, 4 writes): Currencies, Groups, StakeAccounts, Account
+	pub(crate) fn do_init_collator(who: &T::AccountId) -> DispatchResult {
+		T::Currency::mint_into(STAKE_CURRENCY_ID, who, DEFAULT_COLLATOR_STAKE.into())?;
+		T::Rewards::deposit_stake(
+			(T::Domain::get(), STAKE_CURRENCY_ID),
+			who,
+			DEFAULT_COLLATOR_STAKE.into(),
+		)
+	}
+
+	/// Withdraw currently staked amount for target address and immediately burn it.
+	/// Disables receiving rewards onwards.
+	pub(crate) fn do_exit_collator(who: &T::AccountId) -> DispatchResult {
+		let amount = T::Rewards::account_stake((T::Domain::get(), STAKE_CURRENCY_ID), who);
+		T::Rewards::withdraw_stake((T::Domain::get(), STAKE_CURRENCY_ID), who, amount)?;
+		T::Currency::burn_from(STAKE_CURRENCY_ID, who, amount).map(|_| ())
+	}
+
+	/// Apply session changes and distribute rewards.
+	///
+	/// NOTE: Noop if any call fails.
+	fn do_advance_session() {
+		let mut num_joining = 0u32;
+		let mut num_leaving = 0u32;
+
+		transactional::with_storage_layer(|| -> DispatchResult {
+			NextSessionChanges::<T>::try_mutate(|changes| -> DispatchResult {
+				ActiveSessionData::<T>::try_mutate(|session_data| {
+					// Reward collator group of last session
+					let total_collator_reward = session_data
+						.collator_reward
+						.saturating_mul(session_data.collator_count.into())
+						.min(session_data.total_reward);
+					T::Rewards::reward_group(COLLATOR_GROUP_ID, total_collator_reward)?;
+
+					// Handle remaining reward
+					let remaining = session_data
+						.total_reward
+						.saturating_sub(total_collator_reward);
+					if !remaining.is_zero() {
+						let reward = T::Currency::issue(remaining.into());
+						// If configured, assigns reward to Beneficiary, else automatically drops it
+						T::Beneficiary::on_unbalanced(reward);
+					}
+
+					num_joining = changes.collators.inc.len().saturated_into();
+					num_leaving = changes.collators.out.len().saturated_into();
+
+					// Apply collator set changes AFTER rewarding
+					for leaving in changes.collators.out.iter() {
+						Self::do_exit_collator(leaving)?;
+					}
+					for joining in changes.collators.inc.iter() {
+						Self::do_init_collator(joining)?;
+					}
+
+					// Apply session changes
+					session_data.collator_reward = changes
+						.collator_reward
+						.unwrap_or(session_data.collator_reward);
+					session_data.total_reward =
+						changes.total_reward.unwrap_or(session_data.total_reward);
+					session_data.collator_count = changes
+						.collator_count
+						.unwrap_or(session_data.collator_count);
+
+					Self::deposit_event(Event::NewSession {
+						total_reward: session_data.total_reward,
+						collator_reward: session_data.collator_reward,
+						last_changes: mem::take(changes),
+					});
+
+					Ok(())
+				})
+			})
+		})
+		.map_err(|e| {
+			log::error!("Failed to advance block rewards session: {:?}", e);
```

Comment:
If this is the case, I would change the `saturating_ops` above to `ensure_ops`, to avoid "weird invisible behaviors" if those operations "fails". WDYT?
Maybe we should generate some kind of error event, too (not sure if this is the intended behavior of events, but sounds reasonable at least in infallible functions).